contracts/modules/capital/PriceFeedOracle.sol

```diff
@@ -18,40 +18,68 @@ contract PriceFeedOracle is IPriceFeedOracle {
     uint8[] memory _assetDecimals,
     address _safeTracker
   ) {
-    require(_assetAddresses.length > 0, "PriceFeedOracle: asset addresses cannot be empty");
-    require(
-      _assetAddresses.length == _assetAggregators.length &&
-      _assetAggregators.length == _aggregatorTypes.length &&
-      _aggregatorTypes.length == _assetDecimals.length,
-      "PriceFeedOracle: different args length"
-    );
-    require(_safeTracker != address(0), "PriceFeedOracle: safeTracker cannot be zero address");
+    if (_assetAddresses.length == 0) {
+      revert EmptyAssetAddresses();
+    }
+    if (
+      _assetAddresses.length != _assetAggregators.length ||
+      _assetAggregators.length != _aggregatorTypes.length ||
+      _aggregatorTypes.length != _assetDecimals.length
+    ) {
+      revert ArgumentLengthMismatch(
+        _assetAddresses.length,
+        _assetAggregators.length,
+        _aggregatorTypes.length,
+        _assetDecimals.length
+      );
+    }
+    if (_safeTracker == address(0)) {
+      revert ZeroAddress("safeTracker");
+    }
 
     safeTracker = _safeTracker;
     assetsMap[_safeTracker] = AssetInfo(Aggregator(_safeTracker), AggregatorType.ETH, 18);
 
     for (uint i = 0; i < _assetAddresses.length; i++) {
-      require(_assetAddresses[i] != address(0), "PriceFeedOracle: asset address cannot be zero");
-      require(_assetAggregators[i] != address(0), "PriceFeedOracle: aggregator address cannot be zero");
-
-      assetsMap[_assetAddresses[i]] = AssetInfo(
-        Aggregator(_assetAggregators[i]),
-        _aggregatorTypes[i],
-        _assetDecimals[i]
-      );
+      if (_assetAddresses[i] == address(0)) {
+        revert ZeroAddress("assetAddress");
+      }
+      if (_assetAggregators[i] == address(0)) {
+        revert ZeroAddress("aggregator");
+      }
+      if (_assetDecimals[i] == 0) {
+        revert ZeroDecimals(_assetAddresses[i]);
+      }
+
+      Aggregator aggregator = Aggregator(_assetAggregators[i]);
+      uint8 aggregatorDecimals = aggregator.decimals();
+
+      if (_aggregatorTypes[i] != AggregatorType.ETH && _aggregatorTypes[i] != AggregatorType.USD) {
+          revert UnknownAggregatorType(uint8(_aggregatorTypes[i]));
+      }
+      if (_aggregatorTypes[i] == AggregatorType.ETH && aggregatorDecimals != 18) {
+          revert InvalidAggregatorDecimals(_assetAggregators[i], aggregatorDecimals, 18);
+      }
+      if (_aggregatorTypes[i] == AggregatorType.USD && aggregatorDecimals != 8) {
+          revert InvalidAggregatorDecimals(_assetAggregators[i], aggregatorDecimals, 8);
+      }
```

Comment:
If I'd encounter this error it would appear to me that _I have to_ pass 18 or 8 for those. The error should say that the aggregator is incompatible instead, because it returns a different number of decimals than expected.

Reply:
@shark0der rename to `IncompatibleAggregatorDecimals` is ok?
```solidity
  error IncompatibleAggregatorDecimals(address aggregator, uint8 aggregatorDecimals, uint8 expectedDecimals);
```